Adds pose output to DepthSensor and updates DepthSensorToLcmPointCloudMessage to use it.

liangfok opened the pull request:
This is necessary for downstream consumers of the depth sensor's point cloud to convert the point cloud into the world frame.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/5823)
<!-- Reviewable:end -->

liangfok commented:
+@kunimatsu-tri for feature review please. Thanks!

---

Review status: 0 of 8 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5823#-:-Khd08hj3sxbVXlNDGki:bk1y2no)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
Completed the first round.

---

Reviewed 8 of 8 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/sensors/depth_sensor.cc, line 195 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5823#-KhegQ7noUvJa2LIWqOp:-KhegQ7noUvJa2LIWqOq:b-k2z5ki) ([raw file](https://github.com/robotlocomotion/drake/blob/2a8af08f1a239bf57bea7ef86f11e7bfb78c62e6/drake/systems/sensors/depth_sensor.cc#L195)):*
> ```cpp
> 
>   // Evaluates theh output port containing X_WF.
>   const drake::Isometry3<double> X_WF =
> ```

This variable name `X_WF` is a bit confusing since this is stored in `output_x_ws` as it is later.  Then, why`X_WF` instead of `X_WS`?

---

*[drake/systems/sensors/depth_sensor_to_lcm_point_cloud_message.cc, line 59 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5823#-Khekup5ehvn2N3CVqxa:-Khekup5ehvn2N3CVqxb:bkexu9w) ([raw file](https://github.com/robotlocomotion/drake/blob/2a8af08f1a239bf57bea7ef86f11e7bfb78c62e6/drake/systems/sensors/depth_sensor_to_lcm_point_cloud_message.cc#L59)):*
> ```cpp
>   const PoseVector<double> default_X_WS;
>   if (X_WS == nullptr) {
>     X_WS = &default_X_WS;
> ```

The same comment as below.

---

*[drake/systems/sensors/depth_sensor_to_lcm_point_cloud_message.h, line 14 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5823#-KheiD9dw0v6LkMSgTS4:-KheiD9dw0v6LkMSgTS5:b-6j1zss) ([raw file](https://github.com/robotlocomotion/drake/blob/2a8af08f1a239bf57bea7ef86f11e7bfb78c62e6/drake/systems/sensors/depth_sensor_to_lcm_point_cloud_message.h#L14)):*
> ```objc
> /// A DepthSensorToLcmPointCloudMessage takes as input a DepthSensorOutput and
> /// `X_WS`. If the input port containing `X_WS` is not connected, this system
> /// assumes the sensor's frame is coincident with the world frame. It outputs an
> ```

If an user just forgets to connect the input port for `X_WS` by chance, this system outputs wrong point cloud and it may take a while to find for the user from where the undesirable behavior is coming.

Instead, you can design the system to fail (`DRAKE_DEMAND` or exception) if the input port for `X_WS` is not connected. This "fail fast" seems a better choice to me, but what do you think?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5823#-:-KheoYuO5U7_0hP0n23G:b-ou3g7b)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@kunimatsu-tri, thanks for the review. I addressed your comments. PTAL.

---

Review status: 4 of 8 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/sensors/depth_sensor.cc, line 195 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5823#-KhegQ7noUvJa2LIWqOp:-Khi3UrHSqBAcfLtLl0y:bgu9l45) ([raw file](https://github.com/robotlocomotion/drake/blob/2a8af08f1a239bf57bea7ef86f11e7bfb78c62e6/drake/systems/sensors/depth_sensor.cc#L195)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

This variable name `X_WF` is a bit confusing since this is stored in `output_x_ws` as it is later.  Then, why`X_WF` instead of `X_WS`?
</blockquote></details>

Done. It was a copy/paste mistake.

---

*[drake/systems/sensors/depth_sensor_to_lcm_point_cloud_message.cc, line 59 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5823#-Khekup5ehvn2N3CVqxa:-Khi4fMnjj2EBP53dgL2:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/2a8af08f1a239bf57bea7ef86f11e7bfb78c62e6/drake/systems/sensors/depth_sensor_to_lcm_point_cloud_message.cc#L59)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

The same comment as below.
</blockquote></details>

Done.

---

*[drake/systems/sensors/depth_sensor_to_lcm_point_cloud_message.h, line 14 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5823#-KheiD9dw0v6LkMSgTS4:-Khi4gVaA88J0NO6GygK:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/2a8af08f1a239bf57bea7ef86f11e7bfb78c62e6/drake/systems/sensors/depth_sensor_to_lcm_point_cloud_message.h#L14)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

If an user just forgets to connect the input port for `X_WS` by chance, this system outputs wrong point cloud and it may take a while to find for the user from where the undesirable behavior is coming.

Instead, you can design the system to fail (`DRAKE_DEMAND` or exception) if the input port for `X_WS` is not connected. This "fail fast" seems a better choice to me, but what do you think?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5823#-:-Khi58fmwwhRBRfstl0-:btui6hv)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r2.
Review status: 4 of 8 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[drake/systems/sensors/depth_sensor_to_lcm_point_cloud_message.h, line 14 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5823#-KheiD9dw0v6LkMSgTS4:-KhikKbbHqoC60dNQoyD:b-ucfknh) ([raw file](https://github.com/robotlocomotion/drake/blob/2a8af08f1a239bf57bea7ef86f11e7bfb78c62e6/drake/systems/sensors/depth_sensor_to_lcm_point_cloud_message.h#L14)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5823#-:-KhikP4bpYnL6DBfPd67:bnfp4nl)*
<!-- Sent from Reviewable.io -->

liangfok commented:
+@soonho-tri for platform review please. Thanks!

---

Review status: 3 of 8 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5823#-:-Khimqa_sB8H7B5Z7nU5:b-lnsuqk)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@drake-jenkins-bot linux-xenial-clang-ninja-experimental-minimal please.

soonho-tri commented:
Reviewed 2 of 8 files at r1, 2 of 4 files at r3.
Review status: 5 of 8 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5823)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 8 files at r1, 1 of 4 files at r2, 2 of 4 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5823#-:-KhrYqT_v7uFRsat-6JW:bnfp4nl)*
<!-- Sent from Reviewable.io -->
